app/models/course/lecture.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+class Course::Lecture < ActiveRecord::Base
+  acts_as_readable on: :updated_at
+  has_many_attachments on: :content
+
+  belongs_to :course, inverse_of: :lectures
+
+  scope :sorted_by_date, -> { order(start_at: :desc) }
+
+  TIME_BRAINCERT = '%I:%M%p'
+  DATE_ISO = '%Y-%m-%d'
+  BRAINCERT_API_BASE_URL = 'https://api.braincert.com'
+
+  def handle_access_link(user, is_instructor)
+    if (msg = lecture_inactive?).present?
+      return nil, msg
+    end
+    _, error = create_classroom()
+    return [nil, error] if error
+    generate_classroom_link(user, is_instructor)
+  end
+
+  private
+
+  # Entry point for generating virtual classroom link for a Lecture
+  #
+  # @param [User] user The user from whom the request comes
+  # @param [Boolean] is_instructor Whether the user is an instructor
+  #   (has "manage" permission of lecture)
+  # @return [Array] of format [link, errors]
+  def generate_classroom_link(user, is_instructor)
+    res = call_braincert_api('/v2/getclasslaunch',
+                             generate_classroom_link_params(user, is_instructor))
+    res_body = JSON.parse(res.body)
+    if (error = res_body['error']).present?
+      return nil, I18n.t(:'course.lectures.error_generating_link', error: error)
+    else
+      link = res_body['encryptedlaunchurl']
+      update!(instructor_classroom_link: link) if is_instructor
+      [link, nil]
+    end
+  end
+
+  def generate_classroom_link_params(user, is_instructor)
+    lesson_name = title
+    {
+      class_id: classroom_id,
+      userId: user.id,
+      userName: user.name,
+      isTeacher: is_instructor ? '1' : '0',
+      lessonName: lesson_name,
+      courseName: lesson_name
+    }
+  end
+
+  def lecture_inactive?
+    return if currently_active?
+    diff = start_at - Time.zone.now
+    if diff > 0
+      I18n.t(:'course.lectures.lesson_live_in',
+             desc: distance_of_time_in_words(diff))
+    else
+      I18n.t(:'course.lectures.lesson_already_conducted')
+    end
+  end
+
+  def post(url, params)
+    Net::HTTP.post_form URI.parse(url), params
+  end
+
+  def call_braincert_api(action, params)
```

Comment:
Would you prefer to use javascript to do this parts? Feels like javascript-ish code in ruby-style. But if you want to leave it like that I think it's fine, but harder to maintain and harder to handle callbacks.